File: src/python/xss.py
"""CWE-79: Reflected XSS via render_template_string and unsafe Markup."""

from flask import Flask, request, render_template_string, Markup

app = Flask(__name__)


@app.route("/hello")
def hello():
    name = request.args.get("name", "world")
    return render_template_string("<h1>Hello " + name + "!</h1>")


@app.route("/bio")
def bio():
    text = request.args.get("text", "")
    return "<div>" + text + "</div>"


@app.route("/comment")
def comment():
    body = request.args.get("body", "")
    return render_template_string("<p>{{ b|safe }}</p>", b=Markup(body))


if __name__ == "__main__":
    app.run(port=4001)

File: src/javascript/xss.js
// CWE-79: Reflected and stored Cross-Site Scripting.

const express = require('express');
const app = express();

app.get('/hello', (req, res) => {
  const name = req.query.name;
  res.setHeader('Content-Type', 'text/html');
  res.end(`<h1>Hello ${name}!</h1>`);
});

app.get('/profile', (req, res) => {
  const bio = req.query.bio;
  res.send('<div class="bio">' + bio + '</div>');
});

const comments = [];
app.post('/comment', express.urlencoded({ extended: true }), (req, res) => {
  comments.push(req.body.text);
  res.redirect('/comments');
});

app.get('/comments', (req, res) => {
  const html = comments.map(c => `<li>${c}</li>`).join('');
  res.send(`<ul>${html}</ul>`);
});

app.listen(3001);
